src/fifelse.c

```diff
@@ -140,3 +140,184 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
+      error("Length of output value #%d must either be 1 or length of logical condition.", i*2+2);
+    }
+    int64_t amask = len1>1 ? INT64_MAX : 0;
+    switch(TYPEOF(outs)) {
+    case LGLSXP: {
+      const int *restrict pouts = LOGICAL(outs);
+      int *restrict pans = LOGICAL(ans);
+      const int pna = nonna ? LOGICAL(na)[0] : NA_LOGICAL;
+      for (int64_t j=0; j<len2; ++j) {
+        idx = imask ? j : p[j];
+        if (pcons[idx]==1) {
+          pans[idx] = pouts[idx & amask];
+        } else {
+          if (imask) pans[j] = pna;
+          p[l++] = idx;
+        }
+      }
+    } break;
+    case INTSXP: {
+      const int *restrict pouts = INTEGER(outs);
+      int *restrict pans = INTEGER(ans);
+      const int pna = nonna ? INTEGER(na)[0] : NA_INTEGER;
+      for (int64_t j=0; j<len2; ++j) {
+        idx = imask ? j : p[j];
+        if (pcons[idx]==1) {
+          pans[idx] = pouts[idx & amask];
+        } else {
+          if (imask) pans[j] = pna;
+          p[l++] = idx;
+        }
+      }
+    } break;
+    case REALSXP: {
+      const double *restrict pouts = REAL(outs);
+      double *restrict pans = REAL(ans);
+      const double na_double = Rinherits(outs, char_integer64) ? NA_INT64_D : NA_REAL;
+      const double pna = nonna ? REAL(na)[0] : na_double;
+      for (int64_t j=0; j<len2; ++j) {
+        idx = imask ? j : p[j];
+        if (pcons[idx]==1) {
+          pans[idx] = pouts[idx & amask];
+        } else {
+          if (imask) pans[j] = pna;
+          p[l++] = idx;
+        }
+      }
+    } break;
+    case CPLXSXP: {
+      const Rcomplex *restrict pouts = COMPLEX(outs);
+      Rcomplex *restrict pans = COMPLEX(ans);
+      const Rcomplex pna = nonna ? COMPLEX(na)[0] : NA_CPLX;
+      for (int64_t j=0; j<len2; ++j) {
+        idx = imask ? j : p[j];
+        if (pcons[idx]==1) {
+          pans[idx] = pouts[idx & amask];
+        } else {
+          if (imask) pans[j] = pna;
+          p[l++] = idx;
+        }
+      }
+    } break;
+    case STRSXP: {
+      const SEXP *restrict pouts = STRING_PTR(outs);
+      const SEXP pna = nonna ? STRING_PTR(na)[0] : NA_STRING;
+      for (int64_t j=0; j<len2; ++j) {
+        idx = imask ? j : p[j];
+        if (pcons[idx]==1) {
+          SET_STRING_ELT(ans, idx, pouts[idx & amask]);
+        } else {
+          if (imask) SET_STRING_ELT(ans, idx, pna);
+          p[l++] = idx;
+        }
+      }
+    } break;
+    case VECSXP: {
+      const SEXP *restrict pouts = VECTOR_PTR(outs);
+      const SEXP pna = VECTOR_PTR(na)[0];
+      for (int64_t j=0; j<len2; ++j) {
+        idx = imask ? j : p[j];
+        if (pcons[idx]==1) {
+          SET_VECTOR_ELT(ans, idx, pouts[idx & amask]);
+        } else {
+          if (imask && nonna) SET_VECTOR_ELT(ans, idx, pna);
+          p[l++] = idx;
+        }
+      }
+    } break;
+    default:
+      error("Type %s is not supported.", type2char(TYPEOF(outs)));
+    }
+    if (0 == l) break;
```

Comment:
1. I would reorder and write `l == 0` to make it easier to read
2. also please put `break` into new line so codecov can detect if that part of code was checked during tests